source/adapters/level_zero/image.cpp

```diff
@@ -36,75 +595,178 @@ urBindlessImagesUnsampledImageHandleDestroyExp(ur_context_handle_t hContext,
+  auto *UrImage = reinterpret_cast<_ur_image *>(hImageMem);
+  if (!UrImage->RefCount.decrementAndTest())
+    return UR_RESULT_SUCCESS;
+
+  if (UrImage->OwnNativeHandle) {
+    auto ZeResult = ZE_CALL_NOCHECK(zeImageDestroy, (UrImage->ZeImage));
+    // Gracefully handle the case that L0 was already unloaded.
+    if (ZeResult && ZeResult != ZE_RESULT_ERROR_UNINITIALIZED)
+      return ze2urResult(ZeResult);
+  }
+  delete UrImage;
+  return UR_RESULT_SUCCESS;
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urBindlessImagesUnsampledImageCreateExp(
     ur_context_handle_t hContext, ur_device_handle_t hDevice,
     ur_exp_image_mem_handle_t hImageMem, const ur_image_format_t *pImageFormat,
     const ur_image_desc_t *pImageDesc, ur_mem_handle_t *phMem,
     ur_exp_image_handle_t *phImage) {
-  std::ignore = hContext;
-  std::ignore = hDevice;
-  std::ignore = hImageMem;
-  std::ignore = pImageFormat;
-  std::ignore = pImageDesc;
   std::ignore = phMem;
-  std::ignore = phImage;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  std::shared_lock<ur_shared_mutex> Lock(hContext->Mutex);
+
+  ZeStruct<ze_image_desc_t> ZeImageDesc;
+  UR_CALL(ur2zeImageDesc(pImageFormat, pImageDesc, ZeImageDesc));
+
+  _ur_image *UrImage = nullptr;
+
+  ze_memory_allocation_properties_t MemAllocProperties{
+      ZE_STRUCTURE_TYPE_MEMORY_ALLOCATION_PROPERTIES};
+  ZE2UR_CALL(zeMemGetAllocProperties,
+             (hContext->ZeContext, hImageMem, &MemAllocProperties, nullptr));
+  if (MemAllocProperties.type == ZE_MEMORY_TYPE_UNKNOWN) {
+    UrImage = reinterpret_cast<_ur_image *>(hImageMem);
+    if (!isSameImageDesc(&UrImage->ZeImageDesc, &ZeImageDesc)) {
+      ze_image_bindless_exp_desc_t ZeImageBindlessDesc;
+      ZeImageBindlessDesc.stype = ZE_STRUCTURE_TYPE_BINDLESS_IMAGE_EXP_DESC;
+      ZeImageBindlessDesc.pNext = nullptr;
+      ZeImageBindlessDesc.flags = ZE_IMAGE_BINDLESS_EXP_FLAG_BINDLESS;
+      ZeImageDesc.pNext = &ZeImageBindlessDesc;
+      ze_image_handle_t ZeImageView;
+      ZE2UR_CALL(zeImageViewCreateExt,
+                 (hContext->ZeContext, hDevice->ZeDevice, &ZeImageDesc,
+                  UrImage->ZeImage, &ZeImageView));
+      ZE2UR_CALL(zeContextMakeImageResident,
+                 (hContext->ZeContext, hDevice->ZeDevice, ZeImageView));
+      try {
+        UrImage = new _ur_image(hContext, ZeImageView, /*OwnZeMemHandle*/ true);
+        UrImage->ZeImageDesc = ZeImageDesc;
+        *phMem = UrImage;
+      } catch (const std::bad_alloc &) {
+        return UR_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+      } catch (...) {
+        return UR_RESULT_ERROR_UNKNOWN;
+      }
+    } else {
+      *phMem = nullptr;
+    }
+  } else {
+    ze_image_pitched_exp_desc_t ZeImagePitchedDesc;
+    ZeImagePitchedDesc.stype = ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC;
+    ZeImagePitchedDesc.pNext = nullptr;
+    ZeImagePitchedDesc.ptr = hImageMem;
+    ZeImageDesc.pNext = &ZeImagePitchedDesc;
```

Comment:
we need BINDLESS FLAG here too - image from pitched memory must be created with bindless descriptor:

https://spec.oneapi.io/level-zero/latest/core/api.html#ze-image-pitched-exp-desc-t

struct ze_image_pitched_exp_desc_t
**Image descriptor for bindless images** created from pitched allocations

Reply:
So it would be a chain of pNext like your test code:
```
    ze_image_pitched_exp_desc_t pitchedDesc = {};
    pitchedDesc.stype = ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC;
    pitchedDesc.ptr = reinterpret_cast<void *>(0x1234000);

    ze_image_bindless_exp_desc_t bindlessExtDesc = {};
    bindlessExtDesc.stype = ZE_STRUCTURE_TYPE_BINDLESS_IMAGE_EXP_DESC;
    bindlessExtDesc.pNext = &pitchedDesc;
    bindlessExtDesc.flags = ZE_IMAGE_BINDLESS_EXP_FLAG_BINDLESS;
```
ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC is only used for bindless image, right?
Alternatively can we set bindless flag to true when ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC  type is present? This would save us a pNext.
IMHO the chain of two pNext make more sense if ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC has multiple purposes.

`Image descriptor for bindless images created from pitched allocations` also indicates that ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC is only used for bindless image.

Reply:
yes, this has to be a chain.

For now, ZE_STRUCTURE_TYPE_PITCHED_IMAGE_EXP_DESC  is only defined for BINDLESS images - so we need to pass  ze_image_bindless_exp_desc_t   too. 

There could be possible etnesion that will allow creating non-bindless images with pitched pointer - something like OpenCL's:

cl_mem clCreateImageWithProperties(
    cl_context context,
    const cl_mem_properties* properties,
    cl_mem_flags flags,
    const cl_image_format* image_format,
    const cl_image_desc* image_desc,
    void* host_ptr,
    cl_int* errcode_ret);